src/server/coprocessor/endpoint.rs

```diff
@@ -997,7 +993,7 @@ impl<'a> SelectContext<'a> {
             } else {
                 range.get_start().to_vec()
             };
-            let upper_bound = if !self.core.desc_scan && range.has_end() {
+            let upper_bound = if !self.core.desc_scan && !range.get_end().is_empty() {
```

Comment:
I suppose it won't be empty in normal case.
proto3 doesn't have `has_xx()` so I use `is_empty()` as an equivalent replacement.